File: 0320-Generalized Abbreviation/0320-Generalized Abbreviation.cpp
class Solution {
public:
    vector<string> generateAbbreviations(string word) {
        vector<string> result;
        dfs(word, 0, "", false, result);
        return result;
    }
    
private:
    void dfs(string& word, int start, string path, bool prenum, vector<string>& result)
    {
        if (start == word.size())
        {
            result.push_back(path);
            return;
        }
        
        dfs(word, start + 1, path + word[start], false, result);
        if (!prenum)
        {
            for (int i = start; i < word.size(); ++i)
            {
                dfs(word, i + 1, path + to_string(i - start + 1), true, result);
            }
        }
    }
};

File: 0301-0400/0320-Generalized Abbreviation/0320-Generalized Abbreviation.py
class Solution:
    def generateAbbreviations(self, word: str) -> List[str]:
        result = []
        def dfs(start, path, prenum):
            if start == len(word):
                result.append(path)
                return
            dfs(start + 1, path + word[start], False)
            if not prenum:
                for i in range(start, len(word)):
                    dfs(i + 1, path + str(i - start + 1), True)
        
        dfs(0, '', False)
        return result
